File: pro-accrete-starform-stargen/src/com/hipposretribution/system/formstar/star/StarCalc.java
package com.hipposretribution.system.formstar.star;

import java.util.Random;

import com.hipposretribution.system.formstar.SolarConst;
import com.hipposretribution.utils.random.RandomUtil;

/**
 * Class to do the heavy calculations of generating a star
 * 
 * @author zakski
 * 
 */
final class StarCalc {

	/**
	 * Constructor Not Needed
	 */
	private StarCalc() {
	}

	/**
	 * Method to calculate Star's Surface Temperature in kelvin. Equation is derived from the
	 * Stefan–Boltzmann equation.
	 * 
	 * @param stellarLuminosity - star's luminosity in terms of solar luminosity
	 * @param stellarRadius - star's radius in terms of solar radius
	 * @return star's surface temperature in terms of kelvin
	 */
	private static double temperature(double stellarLuminosity, double stellarRadius) {
		return Math.floor(SolarConst.SUN_SURFACE_TEMP * Math.sqrt(Math.sqrt(stellarLuminosity) / stellarRadius));
	}

	/**
	 * Method to approximate the star's radius in terms of solar radius
	 * 
	 * 
	 * Solar radius is a unit of distance used to express the size of stars in astronomy equal to the current
	 * radius of the Sun.
	 * 
	 * @param rand - pseudo-random number generator interface
	 * @param stellarMass - star's mass in terms of solar mass
	 * @return star's radius in terms of solar radius
	 */
	private static double radius(Random rand, double stellarMass) {
		return RandomUtil.variation(rand, Math.pow(stellarMass, 1.0 / 3.0), 0.05);
	}

	/**
	 * Method to calculate the aproximate greenhouse limit of the star
	 * 
	 * TODO: WORK OUT IF THIS FORMULA CANNOT BE IMPROVED, SMELLS LIKE MAGIC NUMBERS!!!
	 * 
	 * @param stellarLuminosity - star's luminosity in terms of solar luminosity
	 * @return inner habitable limit in what I want to say is AU
	 */
	private static double innerHabitableLimit(double radiusEcosphere) {
		return radiusEcosphere * SolarConst.GREENHOUSE_EFFECT_CONST;
	}

	/**
	 * Method to calculate the aproximate ecosphere limit of the star
	 * 
	 * TODO: WORK OUT IF THIS FORMULA CANNOT BE IMPROVED, SMELLS LIKE MAGIC NUMBERS!!!
	 * 
	 * @param stellarLuminosity - star's luminosity in terms of solar luminosity
	 * @return outer habitable limit in what I want to say is AU
	 */
	private static double outerHabitableLimit(double stellarLuminosity) {
		return Math.sqrt(stellarLuminosity);
	}

	/**
	 * Method to generate the star's approximate age based on the main sequence lifespan
	 * 
	 * @param rand - pseudo-random number generator interface
	 * @param mainSeqLife - the main sequence lifespan for upper limit of the generated star' age.
	 * @return the star's generated age
	 */
	private static double age(Random rand, double mainSeqLife) {
		if (mainSeqLife > SolarConst.AGE_OUTER_LIMIT)
			return RandomUtil.nextDouble(rand, SolarConst.AGE_INNER_LIMIT, SolarConst.AGE_OUTER_LIMIT);
		else if (mainSeqLife > SolarConst.AGE_INNER_LIMIT)
			return RandomUtil.nextDouble(rand, SolarConst.AGE_INNER_LIMIT, mainSeqLife);
		else
			return RandomUtil.nextDouble(rand, mainSeqLife / 10, mainSeqLife);
	}

	/**
	 * Method to calculate the star's approximate lifespan on the main sequence
	 * 
	 * @param stellarMass - star's mass in terms of solar mass
	 * @param stellarLuminosity - star's luminosity in terms of solar luminosity
	 * @return star's Main Sequence Lifespan in terms of years
	 */
	private static double mainSequenceLife(double stellarMass, double stellarLuminosity) {
		return SolarConst.SUN_MS_LIFESPAN * (stellarMass / stellarLuminosity);
	}

	/**
	 * Method to calculate the luminosity of a star (in solar luminosity) for the star of a given mass (in
	 * solar masses).
	 * 
	 * TODO: THIS FORMULA IS ONLY ACCURATE FOR STARS WITH A MASS CLOSE TO THAT OF THE SUN'S.IT IS A HOLDOVER
	 * FROM ACCRETE-STARFOM AND NEEDS REPLACING WITH SOMETHING BETTER.!!!
	 * 
	 * The solar luminosity, L☉, is a unit of radiant flux (power emitted in the form of photons)
	 * conventionally used by astronomers to measure the luminosity of stars. One solar luminosity is equal to
	 * the current accepted luminosity of the Sun, which is 3.839×1026 W, or 3.839×1033 erg/s. The value is
	 * slightly higher, 3.939×1026 W (equivalent to 4.382×109 kg/s or 1.9×10−16 M☉/d) if the solar
	 * neutrino radiation is included as well as electromagnetic radiation. The Sun is a weakly variable star
	 * and its luminosity therefore fluctuates.[4] The major fluctuation is the eleven-year solar cycle
	 * (sunspot cycle), which causes a periodic variation of about ±0.1%. Any other variation over the last
	 * 200–300 years is thought to be much smaller than this.
	 * 
	 * @param stellarMass - star's mass in terms of solar mass
	 * @return star's luminosity in terms of solar luminosity
	 */
	private static double luminosity(double stellarMass) {
		double n;

		if (stellarMass < 1.0)
			n = 1.75 * (stellarMass - 0.1) + 3.325;
		else
			n = 0.5 * (2.0 - stellarMass) + 4.4;
		return (Math.pow(stellarMass, n));
	}

	/**
	 * Method to procedurally generate the mass of a star in solar mass.
	 * 
	 * The solar mass (M☉) is a standard unit of mass in astronomy that is used to indicate the masses of
	 * other stars, as well as clusters, nebulae and galaxies. It is equal to the mass of the Sun, about two
	 * nonillion kilograms:
	 * 
	 * @param rand - pseudo-random number generator interface
	 * @return generated mass in terms of solar mass
	 */
	private static double mass(Random rand) {
		return RandomUtil.nextDouble(rand, 0.6, 1.3);
	}

	/**
	 * Method to procedurally generate information for a star
	 * 
	 * @param rand - pseudo-random number generator interface
	 * @return - the generated star
	 */
	static Star formStar(Random rand) {
		Star star = new Star();
		star.stellarMass = mass(rand);
		star.stellarLuminosity = luminosity(star.stellarMass);

		star.mainSeqLife = mainSequenceLife(star.stellarMass, star.stellarLuminosity);
		star.age = StarCalc.age(rand, star.mainSeqLife);

		star.radiusEcosphere = outerHabitableLimit(star.stellarLuminosity);
		star.radiusGreenhouse = innerHabitableLimit(star.radiusEcosphere);

		star.stellarRadius = StarCalc.radius(rand, star.stellarMass);

		star.stellarTemp = StarCalc.temperature(star.stellarLuminosity, star.stellarRadius);

		// TODO: Now get the star's class

		return star;
	}
}

File: src/main/scala/com/szadowsz/stargen/base/starform/calc/StarCalc.scala
package com.szadowsz.stargen.base.starform.calc

import com.szadowsz.stargen.base.starform.constants.StarformConstants
import com.szadowsz.stargen.base.starform.system.bodies.Star

/**
  * @author Zakski : 10/09/2016.
  */
trait StarCalc[S <: Star] {
  this: StarformConstants[S] =>

  /**
    * Method to approximate the luminosity of a star for a given mass. See eq. 1-3 in section "3. Characteristics of The Primary Star" in Extra-solar Planetary
    * Systems: A Microcomputer Simulation.
    *
    * @note The solar luminosity, is a unit of radiant flux (power emitted in the form of photons) conventionally used by astronomers to measure the luminosity
    *       of stars. One solar luminosity is equal to the current accepted luminosity of the Sun, which is 3.839×1026 W, or 3.839×1033 erg/s. The value is
    *       slightly higher, 3.939×1026 W (equivalent to 4.382×109 kg/s or 1.9×10?16 M?/d) if the solar neutrino radiation is included as well as
    *       electromagnetic radiation. The Sun is a weakly variable star and its luminosity therefore fluctuates. The major fluctuation is the eleven-year
    *       solar cycle (sunspot cycle), which causes a periodic variation of about ±0.1%. Any other variation over the last 200–300 years is thought to be
    *       much smaller than this.
    *
    * @see p. 502 Extra-solar Planetary Systems: A Microcomputer Simulation - Martyn J. Fogg
    * @see method luminosity, line 10 in enviro.c - Mat Burdick (accrete)
    * @see method luminosity, line 11 in accrete.c - Keris (starform)
    * @see method luminosity, line 10 in accrete.c - Mat Burdick (starform)
    * @see method LUMINOSITY, line 182 in  Star.java - Carl Burke (starform)
    *
    * @param stellarMass star's mass in terms of solar mass
    * @return star's luminosity in terms of solar luminosity
    */
  def stellarLuminosity(stellarMass: Double): Double = {
    if (stellarMass < 1.0) {
      Math.pow(stellarMass, 1.75 * (stellarMass - 0.1) + 3.325)
    } else {
      Math.pow(stellarMass, 0.5 * (2.0 - stellarMass) + 4.4)
    }
  }

  /**
    * Method to approximate the star's lifespan on the main sequence. eq. 4 in "3. Characteristics of The Primary Star" from "Extra-solar Planetary Systems: A
    * Microcomputer Simulation".
    *
    * @note unit is Byr (1.0E9 years)
    *
    * @see p. 502, Extra-solar Planetary Systems: A Microcomputer Simulation - Martyn J. Fogg
    * @see method generate_stellar_system, line 93 in main.c - Mat Burdick (accrete)
    * @see method getLifetime, line 179 in genstar.cc - Keris (accretion v2)
    * @see method critical_limit, line 283 in accrete.c - Keris (starform)
    * @see method generate_stellar_system, line 85 in starform.c - Mat Burdick (starform)
    * @see method commonConstructor, line 112 in  Star.java - Carl Burke (starform)
    *
    * @param stellarMass - star's mass in terms of solar mass
    * @param stellarLuminosity - star's luminosity in terms of solar luminosity
    * @return star's Main Sequence Lifespan in terms of billon years (Byr)
    */
  def stellarMSLifespan(stellarMass: Double, stellarLuminosity: Double): Double = 10 * (stellarMass / stellarLuminosity)

  /**
    * function to calculate the mean habitable orbit around the star from "3. Characteristics of The Primary Star" in "Extra-solar Planetary Systems:
    * A Microcomputer Simulation".
    *
    * @see eq.5, p.502, Extra-solar Planetary Systems: A Microcomputer Simulation - Martyn J. Fogg
    * @see method generate_stellar_system, line 91 in main.c - Mat Burdick (accrete)
    * @see method GenStar, line 167 in Genstar.c - Andrew Folkins (accretion)
    * @see method GenStar, line 214 in genstar.c - Keris (accretion v1)
    * @see method generate_stellar_system, line 75 in gensys.c - Keris (starform)
    * @see method generate_stellar_system, line 90 in starform.c - Mat Burdick (starform)
    * @see method commonConstructor, line 118 in  Star.java - Carl Burke (starform)
    *
    * @param stellarLuminosity star's luminosity in terms of solar luminosity
    * @return mean habitable radius in AU.
    */
  def ecosphereRadius(stellarLuminosity: Double): Double = Math.sqrt(stellarLuminosity)

  /**
    * function to calculate the closest possible habitable orbit around the star from "3. Characteristics of The Primary Star" in "Extra-solar Planetary
    * Systems: A Microcomputer Simulation".
    *
    * @see eq.6, p.503, Extra-solar Planetary Systems: A Microcomputer Simulation - Martyn J. Fogg
    * @see method generate_stellar_system, line 92 in main.c - Mat Burdick (accrete)
    * @see method GenStar, line 168 in Genstar.c - Andrew Folkins (accretion)
    * @see method GenStar, line 215 in genstar.c - Keris (accretion v1)
    * @see method generate_stellar_system, line 76 in gensys.c - Keris (starform)
    * @see method generate_stellar_system, line 91 in starform.c - Mat Burdick (starform)
    * @see method commonConstructor, line 119 in  Star.java - Carl Burke (starform)
    *
    * @param ecosphereRadius mean habitable radius in AU.
    * @return inner habitable radius in AU.
    */
  def greenhouseRadius(ecosphereRadius: Double): Double = ecosphereRadius * GREENHOUSE_EFFECT_CONST

}
